### PR TITLE
Refactor CorfuServer so that code can be reused in tests

### PR DESCRIPTION
**Pull request by no2chem**
## Overview
Description: This PR refactors CorfuServer so that code for instantiating the server and binding to a port can be reused in tests.

Why should this be merged: Previously, most CorfuServer code was untested because it required running from CorfuServer's ```main``` function. This resulted in a lot of code duplication in tests, and ultimately led to the test code deviating from the CorfuServer code. This change consolidates the test code (especially in ```NettyCommTest```) into CorfuServer, and the test code now re-uses the CorfuServer code. This change should increase coverage significantly and pave the way towards actually using Netty during tests instead of the simulated TestRouter framework.

## Checklist (Definition of Done):
- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

File: infrastructure/src/main/java/org/corfudb/infrastructure/NettyServerRouter.java [hunk 1]
```diff
@@ -4,6 +4,8 @@
 import io.netty.channel.ChannelHandlerContext;
 import io.netty.channel.ChannelInboundHandlerAdapter;
 
+import java.util.EnumMap;
+import java.util.List;
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
```
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/NettyServerRouter.java]: MINOR Remove this unused import 'java.util.concurrent.ConcurrentHashMap'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: infrastructure/src/main/java/org/corfudb/infrastructure/NettyServerRouter.java [hunk 2]
```diff
@@ -97,41 +97,32 @@ protected static void handleUncaughtException(Thread t, @Nonnull Throwable e) {
...
-    public void removeServer(AbstractServer server) {
-        // Iterate through all types of CorfuMsgType, un-registering the handler
-        server.getHandler().getHandledTypes()
-                .forEach(x -> {
-                    handlerMap.remove(x, server);
-                    log.trace("Un-Registered {} to handle messages of type {}", server, x);
-                });
+    @Override
+    @Deprecated
+    public void addServer(AbstractServer server) {
```
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/NettyServerRouter.java]: MAJOR Add the missing @deprecated Javadoc tag. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)
INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: infrastructure/src/main/java/org/corfudb/infrastructure/BaseServer.java [hunk 3]
```diff
@@ -6,6 +6,7 @@
 import java.util.HashMap;
```
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/BaseServer.java]: MINOR Remove this unused import 'java.util.HashMap'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: infrastructure/src/main/java/org/corfudb/infrastructure/BaseServer.java [hunk 4]
```diff
@@ -6,6 +6,7 @@
...
 import java.util.Map;
```
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/BaseServer.java]: MINOR Remove this unused import 'java.util.Map'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: infrastructure/src/main/java/org/corfudb/infrastructure/BaseServer.java [hunk 5]
```diff
@@ -6,6 +6,7 @@
...
 
+import javax.annotation.Nonnull;
 import lombok.Getter;
 import lombok.Setter;
```
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/BaseServer.java]: MINOR Remove this unused import 'lombok.Setter'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java [hunk 6]
```diff
@@ -269,169 +254,231 @@ public static void main(String[] args) {
...
+            .childOption(ChannelOption.TCP_NODELAY, true)
+            .childOption(ChannelOption.ALLOCATOR, PooledByteBufAllocator.DEFAULT);
+    }
+
+    /** Get a "boss" group, which services (accepts) incoming connections.
+     *
+     * @param context       The {@link ServerContext} to use.
+     * @return              A boss group.
+     */
+    public static EventLoopGroup getBossGroup(@Nonnull ServerContext context) {
```
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java]: MAJOR Remove this unused method parameter "context". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1172)

File: infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java [hunk 7]
```diff
@@ -29,32 +34,26 @@
 import java.util.concurrent.Executors;
 import java.util.concurrent.ThreadFactory;
 import java.util.concurrent.atomic.AtomicInteger;
```
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java]: MINOR Remove this unused import 'java.util.concurrent.atomic.AtomicInteger'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java [hunk 8]
```diff
@@ -1,26 +1,31 @@
...
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
 import io.netty.bootstrap.ServerBootstrap;
 import io.netty.buffer.PooledByteBufAllocator;
+import io.netty.channel.Channel;
 import io.netty.channel.ChannelFuture;
 import io.netty.channel.ChannelInitializer;
 import io.netty.channel.ChannelOption;
 import io.netty.channel.EventLoopGroup;
+import io.netty.channel.ServerChannel;
+import io.netty.channel.epoll.Epoll;
```
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java]: MINOR Remove this unused import 'io.netty.channel.epoll.Epoll'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java [hunk 9]
```diff
@@ -269,169 +254,231 @@ public static void main(String[] args) {
...
+            throw new UnrecoverableCorfuInterruptedError(ie);
+        }
+    }
+
+    /** Configure server bootstrap per-channel options, such as TCP options, etc.
+     *
+     * @param context       The {@link ServerContext} to use.
+     * @param bootstrap     The {@link ServerBootstrap} to be configured.
+     */
+    public static void configureBootstrapOptions(@Nonnull ServerContext context,
```
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java]: MAJOR Remove this unused method parameter "context". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1172)

File: infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java [hunk 10]
```diff
@@ -1,26 +1,31 @@
...
 import io.netty.bootstrap.ServerBootstrap;
 import io.netty.buffer.PooledByteBufAllocator;
+import io.netty.channel.Channel;
 import io.netty.channel.ChannelFuture;
 import io.netty.channel.ChannelInitializer;
 import io.netty.channel.ChannelOption;
 import io.netty.channel.EventLoopGroup;
+import io.netty.channel.ServerChannel;
+import io.netty.channel.epoll.Epoll;
+import io.netty.channel.epoll.EpollEventLoopGroup;
```
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java]: MINOR Remove this unused import 'io.netty.channel.epoll.EpollEventLoopGroup'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java [hunk 11]
```diff
@@ -165,20 +164,8 @@
...
-    private static SequencerServer sequencerServer;
-    @Getter
-    private static LayoutServer layoutServer;
-    @Getter
-    private static LogUnitServer logUnitServer;
-    @Getter
-    private static ManagementServer managementServer;
-    private static NettyServerRouter router;
+
     private static ServerContext serverContext;
```
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java]: MINOR Remove the "serverContext" field and declare it as a local variable in the relevant methods. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1450)

**Comment by corfu-repo** [comment]
SonarQube analysis reported 12 issues
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 8 minor
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: infrastructure/src/main/java/org/corfudb/infrastructure/NettyServerRouter.java (hunk 1 above)
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/NettyServerRouter.java]: MINOR Remove this unused import 'java.util.concurrent.ConcurrentHashMap'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: infrastructure/src/main/java/org/corfudb/infrastructure/NettyServerRouter.java (hunk 2 above)
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/NettyServerRouter.java]: MAJOR Add the missing @deprecated Javadoc tag. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)
INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: infrastructure/src/main/java/org/corfudb/infrastructure/BaseServer.java (hunk 3 above)
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/BaseServer.java]: MINOR Remove this unused import 'java.util.HashMap'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: infrastructure/src/main/java/org/corfudb/infrastructure/BaseServer.java (hunk 4 above)
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/BaseServer.java]: MINOR Remove this unused import 'java.util.Map'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: infrastructure/src/main/java/org/corfudb/infrastructure/BaseServer.java (hunk 5 above)
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/BaseServer.java]: MINOR Remove this unused import 'lombok.Setter'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java [hunk 12]
```diff
@@ -269,169 +250,231 @@ public static void main(String[] args) {
...
+            .childOption(ChannelOption.TCP_NODELAY, true)
+            .childOption(ChannelOption.ALLOCATOR, PooledByteBufAllocator.DEFAULT);
+    }
+
+    /** Get a "boss" group, which services (accepts) incoming connections.
+     *
+     * @param context       The {@link ServerContext} to use.
+     * @return              A boss group.
+     */
+    public static EventLoopGroup getBossGroup(@Nonnull ServerContext context) {
```
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java]: MAJOR Remove this unused method parameter "context". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1172)

File: infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java [hunk 13]
```diff
@@ -269,169 +250,231 @@ public static void main(String[] args) {
...
+            throw new UnrecoverableCorfuInterruptedError(ie);
+        }
+    }
+
+    /** Configure server bootstrap per-channel options, such as TCP options, etc.
+     *
+     * @param context       The {@link ServerContext} to use.
+     * @param bootstrap     The {@link ServerBootstrap} to be configured.
+     */
+    public static void configureBootstrapOptions(@Nonnull ServerContext context,
```
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java]: MAJOR Remove this unused method parameter "context". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1172)

File: infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java [hunk 14]
```diff
@@ -165,20 +160,8 @@
...
-    private static SequencerServer sequencerServer;
-    @Getter
-    private static LayoutServer layoutServer;
-    @Getter
-    private static LogUnitServer logUnitServer;
-    @Getter
-    private static ManagementServer managementServer;
-    private static NettyServerRouter router;
+
     private static ServerContext serverContext;
```
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java]: MINOR Remove the "serverContext" field and declare it as a local variable in the relevant methods. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1450)

**Comment by corfu-repo** [comment]
SonarQube analysis reported 9 issues
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 5 minor
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **9e19914**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #1061 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/SERF0aCBvFXz7S05KuXBWWE7znGHCDW6)

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **4fa64b8**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #1061 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/DqSF62Wz48YvCbVaY5143fmcSXsQzJRX)

File: test/src/test/java/org/corfudb/runtime/clients/NettyCommTest.java [hunk 15]
```diff
@@ -448,6 +451,8 @@ void runWithBaseServer(NettyServerDataConstructor nsdc,
             ncr.start();
             actionFn.runTest(ncr, d);
         } catch (Exception ex) {
+            System.out.println("oops");
```
Comment (medhavidhawan) [inline test/src/test/java/org/corfudb/runtime/clients/NettyCommTest.java]: System out ?

File: test/src/test/java/org/corfudb/runtime/clients/NettyCommTest.java [hunk 16]
```diff
@@ -448,6 +451,8 @@ void runWithBaseServer(NettyServerDataConstructor nsdc,
...
+            ex.printStackTrace();
```
Comment (medhavidhawan) [inline test/src/test/java/org/corfudb/runtime/clients/NettyCommTest.java]: e.printStackTrace ?

File: infrastructure/src/main/java/org/corfudb/infrastructure/NettyServerRouter.java [hunk 17]
```diff
@@ -97,39 +96,32 @@ protected static void handleUncaughtException(Thread t, @Nonnull Throwable e) {
...
```
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/NettyServerRouter.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java [hunk 18]
```diff
@@ -264,172 +252,231 @@ public static void main(String[] args) {
...
+            .childOption(ChannelOption.TCP_NODELAY, true)
+            .childOption(ChannelOption.ALLOCATOR, PooledByteBufAllocator.DEFAULT);
+    }
+
+    /** Get a "boss" group, which services (accepts) incoming connections.
+     *
+     * @param context       The {@link ServerContext} to use.
+     * @return              A boss group.
+     */
+    public static EventLoopGroup getBossGroup(@Nonnull ServerContext context) {
```
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java]: MAJOR Remove this unused method parameter "context". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1172)

File: infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java [hunk 19]
```diff
@@ -264,172 +252,231 @@ public static void main(String[] args) {
...
+            throw new UnrecoverableCorfuInterruptedError(ie);
+        }
+    }
+
+    /** Configure server bootstrap per-channel options, such as TCP options, etc.
+     *
+     * @param context       The {@link ServerContext} to use.
+     * @param bootstrap     The {@link ServerBootstrap} to be configured.
+     */
+    public static void configureBootstrapOptions(@Nonnull ServerContext context,
```
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java]: MAJOR Remove this unused method parameter "context". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1172)

File: infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java [hunk 20]
```diff
@@ -9,21 +9,23 @@
...
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
 import io.netty.bootstrap.ServerBootstrap;
 import io.netty.buffer.PooledByteBufAllocator;
+import io.netty.channel.Channel;
 import io.netty.channel.ChannelFuture;
 import io.netty.channel.ChannelInitializer;
 import io.netty.channel.ChannelOption;
 import io.netty.channel.EventLoopGroup;
+import io.netty.channel.ServerChannel;
+import io.netty.channel.epoll.Epoll;
```
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java]: MINOR Remove this unused import 'io.netty.channel.epoll.Epoll'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

File: infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java [hunk 21]
```diff
@@ -9,21 +9,23 @@
...
+import io.netty.channel.epoll.EpollEventLoopGroup;
```
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java]: MINOR Remove this unused import 'io.netty.channel.epoll.EpollEventLoopGroup'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by corfu-repo** [comment]
SonarQube analysis reported 7 issues
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 3 minor
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [ServerContext.java#L85](https://github.com/CorfuDB/CorfuDB/blob/43839dbe0320c553d930f16fcb0c4824a4f18dd5/infrastructure/src/main/java/org/corfudb/infrastructure/ServerContext.java#L85): Remove this useless assignment; "serverRouter" already holds the assigned value along all execution paths. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS4165)
1. ![MINOR][MINOR] [BaseServer.java#L7](https://github.com/CorfuDB/CorfuDB/blob/43839dbe0320c553d930f16fcb0c4824a4f18dd5/infrastructure/src/main/java/org/corfudb/infrastructure/BaseServer.java#L7): Remove this unused import 'java.util.Map'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: infrastructure/src/main/java/org/corfudb/infrastructure/NettyServerRouter.java (hunk 17 above)
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/NettyServerRouter.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java [hunk 22]
```diff
@@ -265,171 +250,252 @@ public static void main(String[] args) {
...
+            .childOption(ChannelOption.TCP_NODELAY, true)
+            .childOption(ChannelOption.ALLOCATOR, PooledByteBufAllocator.DEFAULT);
+    }
+
+    /** Get a "boss" group, which services (accepts) incoming connections.
+     *
+     * @param context       The {@link ServerContext} to use.
+     * @return              A boss group.
+     */
+    public static EventLoopGroup getBossGroup(@Nonnull ServerContext context) {
```
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java]: MAJOR Remove this unused method parameter "context". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1172)

File: infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java [hunk 23]
```diff
@@ -265,171 +250,252 @@ public static void main(String[] args) {
...
+            throw new UnrecoverableCorfuInterruptedError(ie);
+        }
+    }
+
+    /** Configure server bootstrap per-channel options, such as TCP options, etc.
+     *
+     * @param context       The {@link ServerContext} to use.
+     * @param bootstrap     The {@link ServerBootstrap} to be configured.
+     */
+    public static void configureBootstrapOptions(@Nonnull ServerContext context,
```
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java]: MAJOR Remove this unused method parameter "context". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1172)

**Comment by corfu-repo** [comment]
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 1 minor
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [ServerContext.java#L85](https://github.com/CorfuDB/CorfuDB/blob/c92b187cbcd7864ba2e32681c875df4f9d104d43/infrastructure/src/main/java/org/corfudb/infrastructure/ServerContext.java#L85): Remove this useless assignment; "serverRouter" already holds the assigned value along all execution paths. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS4165)
1. ![MINOR][MINOR] [BaseServer.java#L7](https://github.com/CorfuDB/CorfuDB/blob/c92b187cbcd7864ba2e32681c875df4f9d104d43/infrastructure/src/main/java/org/corfudb/infrastructure/BaseServer.java#L7): Remove this unused import 'java.util.Map'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **c92b187**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #1061 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/xdV6NzdAHrW5V1RpmX5n04mby02zqKNs)

File: infrastructure/src/main/java/org/corfudb/infrastructure/NettyServerRouter.java [hunk 24]
```diff
@@ -97,39 +96,32 @@ protected static void handleUncaughtException(Thread t, @Nonnull Throwable e) {
...
-    public NettyServerRouter(Map<String, Object> opts) {
-        handlerMap = new ConcurrentHashMap<>();
-    }
+    /** The {@link AbstractServer}s this {@link NettyServerRouter} routes messages for. */
+    @Getter
+    final List<AbstractServer> servers;
 
-    /**
-     * Add a new netty server handler to the router.
+    /** Construct an new {@link NettyServerRouter}.
```
Comment (annym) [inline infrastructure/src/main/java/org/corfudb/infrastructure/NettyServerRouter.java]: typo: an -> a

File: infrastructure/src/main/java/org/corfudb/infrastructure/ServerContext.java [hunk 25]
```diff
@@ -76,13 +77,11 @@
...
      * @param serverConfig map of configuration strings to objects
-     * @param serverRouter server router
      */
-    public ServerContext(Map<String, Object> serverConfig, IServerRouter serverRouter) {
+    public ServerContext(Map<String, Object> serverConfig) {
         this.serverConfig = serverConfig;
         this.dataStore = new DataStore(serverConfig);
         generateNodeId();
-        this.serverRouter = serverRouter;
         this.failureDetectorPolicy = new PeriodicPollPolicy();
```
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/ServerContext.java]: MAJOR Remove this useless assignment; "serverRouter" already holds the assigned value along all execution paths. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS4165)

File: infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java [hunk 26]
```diff
@@ -265,171 +250,252 @@ public static void main(String[] args) {
...
+
+
+    /** Obtain a {@link ChannelInitializer} which initializes the channel pipeline
+     *  for a new {@link ServerChannel}.
+     *
+     * @param context   The {@link ServerContext} to use.
+     * @param router    The {@link NettyServerRouter} to initialize the channel with.
+     * @return          A {@link ChannelInitializer} to intialize the channel.
+     */
+    private static ChannelInitializer getServerChannelInitializer(@Nonnull ServerContext context,
```
Comment (corfu-repo) [inline infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java]: CRITICAL Refactor this method to reduce its Cognitive Complexity from 17 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)

**Comment by corfu-repo** [comment]
SonarQube analysis reported 6 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 1 minor
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [BaseServer.java#L7](https://github.com/CorfuDB/CorfuDB/blob/313a837cc54f0bcd7571d6fa0a45eeb380bb47b4/infrastructure/src/main/java/org/corfudb/infrastructure/BaseServer.java#L7): Remove this unused import 'java.util.Map'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**